lib/Transforms/LLZKPCLLoweringPass.cpp

```diff
@@ -0,0 +1,415 @@
+
+using namespace mlir;
+using namespace llzk;
+using namespace llzk::cast;
+using namespace llzk::boolean;
+using namespace llzk::constrain;
+using namespace llzk::felt;
+using namespace llzk::function;
+using namespace llzk::component;
+using namespace llzk::constrain;
+
+namespace {
+
+static FailureOr<Value> lookup(Value v, llvm::DenseMap<Value, Value> &m, Operation *onError) {
+  if (auto it = m.find(v); it != m.end()) {
+    return it->second;
+  }
+  return onError->emitError("missing operand mapping");
+}
+
+static void rememberResult(Value from, Value to, llvm::DenseMap<Value, Value> &m) {
+  (void)m.try_emplace(from, to);
+}
+
+// Convert binary LLZK op to corresponding binary PCL op
+template <typename SrcBinOp, typename DstBinOp>
+static LogicalResult
+lowerBinaryLike(OpBuilder &b, SrcBinOp src, llvm::DenseMap<Value, Value> &mapping) {
+  auto loc = src.getLoc();
+  auto op = src.getOperation();
+  auto lhs = lookup(src.getLhs(), mapping, op);
+  if (failed(lhs)) {
+    return failure();
+  }
+  auto rhs = lookup(src.getRhs(), mapping, op);
+  if (failed(rhs)) {
+    return failure();
+  }
+
+  auto dst = b.create<DstBinOp>(loc, *lhs, *rhs);
+  rememberResult(src.getResult(), dst.getRes(), mapping);
+  return success();
+}
+
+static LogicalResult
+lowerConst(OpBuilder &b, FeltConstantOp cst, llvm::DenseMap<Value, Value> &mapping) {
+  auto attr = pcl::FeltAttr::get(b.getContext(), cst.getValue());
+  auto dst = b.create<pcl::ConstOp>(cst.getLoc(), attr);
+  rememberResult(cst.getResult(), dst.getRes(), mapping);
+  return success();
+}
+
+class PCLLoweringPass : public llzk::impl::PCLLoweringPassBase<PCLLoweringPass> {
+
+private:
+  void getDependentDialects(mlir::DialectRegistry &registry) const override {
+    registry.insert<pcl::PCLDialect, func::FuncDialect>();
+  }
+
+  /// The translation only works now on LLZK structs where all the fields are felts.
+  LogicalResult validateStruct(StructDefOp structDef) {
+    for (auto field : structDef.getFieldDefs()) {
+      auto fieldType = field.getType();
+      if (!llvm::dyn_cast<FeltType>(fieldType)) {
+        return field.emitError() << "Field must be felt type. Found " << fieldType
+                                 << " for field: " << field.getName();
+      }
+    }
+    return success();
+  }
+
+  /// Emit assertions for an equality `lhs == rhs`, with fast paths when one side
+  /// is a boolean and the other side is a constant {0,1}.
+  ///
+  /// Cases handled:
+  ///   - bool == 1  → assert(bool)
+  ///   - 1 == bool  → assert(bool)
+  ///   - bool == 0  → assert(!bool)
+  ///   - 0 == bool  → assert(!bool)
+  ///   - otherwise  → assert(lhs == rhs)
+  ///
+  /// Returns success after emitting IR.
+  static LogicalResult
+  emitAssertEqOptimized(OpBuilder &b, Location loc, Value lhsVal, Value rhsVal) {
+    // --- Small helpers --------------------------------------------------------
+    auto isBool = [](mlir::Value v) { return llvm::isa<pcl::BoolType>(v.getType()); };
```

Comment:
have `using namespace mlir;` above
```suggestion
    auto isBool = [](Value v) { return llvm::isa<pcl::BoolType>(v.getType()); };
```